File: internal/service/schedule_service.go
package service

import (
	"errors"
	"teaching-management/internal/model"
	"teaching-management/internal/repository"

	"github.com/google/uuid"
)

type ScheduleService struct {
	scheduleRepo *repository.ScheduleRepository
	studentRepo  *repository.StudentRepository
	subjectRepo  *repository.SubjectRepository
}

func NewScheduleService(
	scheduleRepo *repository.ScheduleRepository,
	studentRepo *repository.StudentRepository,
	subjectRepo *repository.SubjectRepository,
) *ScheduleService {
	return &ScheduleService{
		scheduleRepo: scheduleRepo,
		studentRepo:  studentRepo,
		subjectRepo:  subjectRepo,
	}
}

func (s *ScheduleService) CreateSchedule(userID uuid.UUID, req *model.Schedule) (*model.Schedule, error) {
	student, err := s.studentRepo.GetByID(req.StudentID, userID)
	if err != nil {
		return nil, err
	}
	if student == nil {
		return nil, errors.New("student not found or unauthorized")
	}

	subject, err := s.subjectRepo.GetByID(req.SubjectID)
	if err != nil {
		return nil, err
	}
	if subject == nil || subject.UserID != userID {
		return nil, errors.New("subject not found or unauthorized")
	}

	if req.DayOfWeek < 0 || req.DayOfWeek > 6 {
		return nil, errors.New("day of week must be between 0 (Sunday) and 6 (Saturday)")
	}

	schedule := &model.Schedule{
		StudentID: req.StudentID,
		SubjectID: req.SubjectID,
		DayOfWeek: req.DayOfWeek,
		StartTime: req.StartTime,
		EndTime:   req.EndTime,
		IsActive:  true,
	}

	err = s.scheduleRepo.Create(schedule)
	if err != nil {
		return nil, err
	}

	return schedule, nil
}

func (s *ScheduleService) GetSchedules(userID uuid.UUID) ([]model.ScheduleResponse, error) {
	return s.scheduleRepo.GetAll(userID)
}

func (s *ScheduleService) UpdateSchedule(userID uuid.UUID, id uuid.UUID, req *model.Schedule) error {
	existing, err := s.scheduleRepo.GetByID(id)
	if err != nil {
		return err
	}
	if existing == nil {
		return errors.New("schedule not found")
	}

	student, err := s.studentRepo.GetByID(existing.StudentID, userID)
	if err != nil {
		return err
	}
	if student == nil {
		return errors.New("unauthorized access to schedule")
	}

	if req.StudentID != existing.StudentID {
		newStudent, err := s.studentRepo.GetByID(req.StudentID, userID)
		if err != nil {
			return err
		}
		if newStudent == nil {
			return errors.New("new student not found or unauthorized")
		}
	}

	if req.SubjectID != existing.SubjectID {
		newSubject, err := s.subjectRepo.GetByID(req.SubjectID)
		if err != nil {
			return err
		}
		if newSubject == nil || newSubject.UserID != userID {
			return errors.New("new subject not found or unauthorized")
		}
	}

	existing.StudentID = req.StudentID
	existing.SubjectID = req.SubjectID
	existing.DayOfWeek = req.DayOfWeek
	existing.StartTime = req.StartTime
	existing.EndTime = req.EndTime
	existing.IsActive = req.IsActive

	return s.scheduleRepo.Update(existing)
}

func (s *ScheduleService) DeleteSchedule(userID uuid.UUID, id uuid.UUID) error {
	existing, err := s.scheduleRepo.GetByID(id)
	if err != nil {
		return err
	}
	if existing == nil {
		return errors.New("schedule not found")
	}

	student, err := s.studentRepo.GetByID(existing.StudentID, userID)
	if err != nil {
		return err
	}
	if student == nil {
		return errors.New("unauthorized access to schedule")
	}

	return s.scheduleRepo.Delete(id)
}

File: src/services/schedule.service.ts
import {
  ScheduleRepository,
  StudentRepository,
  SubjectRepository
} from '../repositories/interfaces.js';
import { Schedule, ScheduleResponse } from '../types/index.js';

export class ScheduleService {
  constructor(
    private scheduleRepo: ScheduleRepository,
    private studentRepo: StudentRepository,
    private subjectRepo: SubjectRepository
  ) {}

  async createSchedule(
    userId: string,
    req: Omit<Schedule, 'id' | 'created_at' | 'updated_at'>
  ): Promise<Schedule> {
    const student = await this.studentRepo.getById(req.student_id, userId);
    if (!student) {
      throw new Error('student not found or unauthorized');
    }

    const subject = await this.subjectRepo.getById(req.subject_id);
    if (!subject || subject.user_id !== userId) {
      throw new Error('subject not found or unauthorized');
    }

    if (req.day_of_week < 0 || req.day_of_week > 6) {
      throw new Error('day of week must be between 0 (Sunday) and 6 (Saturday)');
    }

    return this.scheduleRepo.create({
      student_id: req.student_id,
      subject_id: req.subject_id,
      day_of_week: req.day_of_week,
      start_time: req.start_time,
      end_time: req.end_time,
      is_active: req.is_active !== undefined ? req.is_active : true
    });
  }

  async getSchedules(userId: string): Promise<ScheduleResponse[]> {
    return this.scheduleRepo.getAll(userId);
  }

  async updateSchedule(userId: string, id: string, req: Schedule): Promise<void> {
    const existing = await this.scheduleRepo.getById(id);
    if (!existing) {
      throw new Error('schedule not found');
    }

    const student = await this.studentRepo.getById(existing.student_id, userId);
    if (!student) {
      throw new Error('unauthorized access to schedule');
    }

    if (req.student_id !== existing.student_id) {
      const newStudent = await this.studentRepo.getById(req.student_id, userId);
      if (!newStudent) {
        throw new Error('new student not found or unauthorized');
      }
    }

    if (req.subject_id !== existing.subject_id) {
      const newSubject = await this.subjectRepo.getById(req.subject_id);
      if (!newSubject || newSubject.user_id !== userId) {
        throw new Error('new subject not found or unauthorized');
      }
    }

    if (req.day_of_week < 0 || req.day_of_week > 6) {
      throw new Error('day of week must be between 0 (Sunday) and 6 (Saturday)');
    }

    await this.scheduleRepo.update({
      ...existing,
      student_id: req.student_id,
      subject_id: req.subject_id,
      day_of_week: req.day_of_week,
      start_time: req.start_time,
      end_time: req.end_time,
      is_active: req.is_active
    });
  }

  async deleteSchedule(userId: string, id: string): Promise<void> {
    const existing = await this.scheduleRepo.getById(id);
    if (!existing) {
      throw new Error('schedule not found');
    }

    const student = await this.studentRepo.getById(existing.student_id, userId);
    if (!student) {
      throw new Error('unauthorized access to schedule');
    }

    await this.scheduleRepo.delete(id);
  }
}
